src/jgikbase/idmapping/service/mapper_service.py

```diff
@@ -130,16 +175,62 @@ def _get_object_id_list_from_json(request) -> List[str]:
     return ids
 
 
-def create_app(builder: IDMappingBuilder=IDMappingBuilder()):
+class JSONFlaskLogFormatter(Formatter):
+    """ A JSON formatter for service logs. """
+
+    def __init__(self, service_name):
+        super().__init__()
+        self.service_name = service_name
+
+    def format(self, record):
+        log = {'service': self.service_name,
+               'level': record.levelname,
+               'time': epoch_ms(),
+               'source': record.name,
+               'ip': flask_req_global.ip,
+               'method': flask_req_global.method,
+               'callid': flask_req_global.req_id,
+               'msg': record.getMessage()
+               }
+        # https://docs.python.org/3.6/library/sys.html#sys.exc_info
+        if record.exc_info and record.exc_info != (None, None, None):
+            log['excep'] = _format_exception(record.exc_info[1])
+        return json.dumps(log)
+
+
+def _configure_loggers(logstream: IO[str]=None):
+    # make some of this configurable if needed
+    handler = StreamHandler(logstream)
+    handler.setFormatter(JSONFlaskLogFormatter('IDMappingService'))
+    logging.getLogger().addHandler(handler)
+    logging.getLogger().setLevel('INFO')
+    logging.getLogger('werkzeug').setLevel('WARNING')
+    logging.getLogger('flask.app').setLevel('WARNING')
+
+
+def create_app(builder: IDMappingBuilder=IDMappingBuilder(), logstream: IO[str]=None):
     """ Create the flask app. """
+    _configure_loggers(logstream)
     app = Flask(__name__)
     app.url_map.strict_slashes = False  # otherwise GET /loc/ won't match GET /loc
     app.config[_APP] = builder.build_id_mapping_system()
+    app.config[_IGNORE_IP_HEADERS] = builder.get_cfg().ignore_ip_headers
 
     @app.before_request
-    def add_request_id():
-        flask_req_global.req_id = str(random.randrange(10000000000000000)).zfill(16)  # nosec
+    def preprocess_request():
         # bandit doesn't like random for crypo purposes, but we're not doing that here
+        flask_req_global.req_id = str(random.randrange(10000000000000000)).zfill(16)  # nosec
```

Comment:
how about using `uuid`?

Reply:
This is the format that almost all the other services use, so I stuck with it. UUIDs would be fine, but it doesn't really matter IMO.